### PR TITLE
Implements vehicle damage based on kinetic energy

### PR DESCRIPTION
**Pull request by nkoenig**
This is an alternative to PR #858. I'm using the kinetic energy formula, `KE = 0.5 * M * v * v`, to calculate a kinetic energy threshold. If the change in KE between iterations passes this threshold, then a crash has occurred and the robot will be disabled. The value of `v` is set to the velocity achieved by falling a certain distance. Currently this distance is 0.077m.

Requires: https://github.com/ignitionrobotics/ign-gazebo/pull/728

Signed-off-by: Nate Koenig <nate@openrobotics.org>

### CODE REVIEW

**Review by Lobotuerk — APPROVED** [review]

File: subt_ign/src/GameLogicPlugin.cc [hunk 1]
```diff
@@ -1012,6 +1040,163 @@ void GameLogicPluginPrivate::OnEvent(const ignition::msgs::Pose &_msg)
...
+                 _ecm.CreateComponent(model->Data(), components::HaltMotion());
+              }
+            }
+          }
+          return true;
+        });
+  }
+
+  // Check for crashes
+  for (auto &ke : this->dataPtr->keInfo)
```
Comment (Lobotuerk) [inline subt_ign/src/GameLogicPlugin.cc:1154]: Should not this part be inside an `else` after `(!this->dataPtr->started)` to make sure no crash happens before starting, like on spawn?

File: subt_ign/src/GameLogicPlugin.cc [hunk 2]
```diff
@@ -1012,6 +1049,163 @@ void GameLogicPluginPrivate::OnEvent(const ignition::msgs::Pose &_msg)
...
+          localSimTime.set_sec(sec);
+          localSimTime.set_nsec(nsec);
+
+          std::lock_guard<std::mutex> lock(this->dataPtr->eventCounterMutex);
+          // _resp.set_report_status("scoring finished");
+          std::ostringstream stream;
+          stream
+            << "- event:\n"
+            << "  id: " << this->dataPtr->eventCounter << "\n"
+            << "  type: kinetic\n"
```
Comment (angelacmaio) [inline subt_ign/src/GameLogicPlugin.cc:1188]: Can we use "collision" as the event name instead?

File: subt_ign/src/GameLogicPlugin.cc [hunk 3]
```diff
@@ -1012,6 +1049,163 @@ void GameLogicPluginPrivate::OnEvent(const ignition::msgs::Pose &_msg)
...
+        {
+          int64_t sec, nsec;
+          std::tie(sec, nsec) =
+            ignition::math::durationToSecNsec(_info.simTime);
+          ignition::msgs::Time localSimTime;
+          localSimTime.set_sec(sec);
+          localSimTime.set_nsec(nsec);
+
+          std::lock_guard<std::mutex> lock(this->dataPtr->eventCounterMutex);
+          // _resp.set_report_status("scoring finished");
```
Comment (angelacmaio) [inline subt_ign/src/GameLogicPlugin.cc:1183]: Comment can be removed

**Comment by peci1** [comment]
#920 inadvarently merged this

**Comment by nkoenig** [comment]
> #920 inadvarently merged this

Thanks. Reverted in d83b71b.

File: subt_ign/src/GameLogicPlugin.cc (hunk 2 above)
Comment (nkoenig) [reply subt_ign/src/GameLogicPlugin.cc]: Done in 56f62bcb12661f5e8338fa7f9abdadaa22775b59

File: subt_ign/src/GameLogicPlugin.cc (hunk 3 above)
Comment (nkoenig) [reply subt_ign/src/GameLogicPlugin.cc]: Removed in 56f62bcb12661f5e8338fa7f9abdadaa22775b59

File: subt_ign/src/GameLogicPlugin.cc [hunk 4]
```diff
@@ -1012,6 +1049,174 @@ void GameLogicPluginPrivate::OnEvent(const ignition::msgs::Pose &_msg)
...
+            {
+              // This sets the kinetic energy threshold for the robot. It is
+              // based on the kinetic energy formula KE = 0.5 * M * v * v
+              //
+              // M is the mass of the robot.
+              // v is velocity of the robot. We are using the velocity acheived
+              // by falling from a height.
+              this->dataPtr->keInfo[model->Data()].kineticEnergyThreshold =
+                0.5 * mass * std::pow(
+                    sqrt((2 * this->dataPtr->keHeight) / 9.8) * 9.8, 2);
```
Comment (acschang) [inline subt_ign/src/GameLogicPlugin.cc:1143]: This ends up evaluating equivocally however the velocity can be simplified from:
`sqrt((2 * this->dataPtr->keHeight) / 9.8) * 9.8`
to:
`sqrt((2 * this->dataPtr->keHeight) * 9.8)`
which would omit the cost of one division.

Comment (nkoenig) [reply subt_ign/src/GameLogicPlugin.cc]: Done in 102aa622dc0ad1b5eeef35903dc42774ed889f7c

File: subt_ign/include/subt_ign/RobotPlatformTypes.hh [hunk 5]
```diff
@@ -23,36 +23,36 @@ const std::map<std::string, double> robotPlatformTypes = {
...
+  {"SHAFTER", 8}, // UAV, no prop guards
   {"SPOT", 1},
   {"TEAMBASE", 1},
   {"X1", 1},
   {"X2", 1},
-  {"X3", 7.2},
-  {"X4", 7.2},
-  {"X500", 14}
+  {"X3", 8}, // UAV, no prop guards
+  {"X6", 8}, // UAV, no prop guards
```
Comment (acschang) [inline subt_ign/include/subt_ign/RobotPlatformTypes.hh:54]: @nkoenig was changing this from `X4` to `X6` intended?

Comment (acschang) [reply subt_ign/include/subt_ign/RobotPlatformTypes.hh]: Fixed in https://github.com/osrf/subt/pull/916/commits/3a7692b8615143a021251ee2efe1ee2fc15b897d

**Pull request merged**